### ISSUE TITLE
Add Grafana 7 cell style to show sanitized HTML 

### ISSUE THREAD

**Issue by maemigh** [grafana/grafana#24482]
**What happened**:
When adding new tables to Grafana 7, many options are missing. "Column Styles" is missing from the panel options. Dashboards migrated from Grafana 6 still include all the options.

**What you expected to happen**:
All options should be available (such as column type, sanitize html, etc) somewhere. Particularly looking for sanitize html.

**How to reproduce it (as minimally and precisely as possible)**:
Create a new table in Grafana 7

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.0-beta3
- Data source type & version: MySQL
- OS Grafana is installed on: RedHat 7.8
- User OS & Browser: Chrome 81.0.4044.129
- Grafana plugins:
- Others:

**Comment by torkelo** [comment]
We have column styles (available via the overrides tab), to set options per column. 

When you say your missing sanitize, do you want to show HTML in a cell?

**Comment by maemigh** [comment]
@torkelo Yes, exactly. My Grafana 6 tables show the formatted HTML (with sanitize HTML turned on), but there doesn't seem to be an option to enable this on Grafana 7 tables. The column style setting of "String" in the overrides does not seem to exist.

**Comment by torkelo** [comment]
yea, we have not added support for this cell display mode yet. 

Just curious what is the value you want to render as html? An anchor link? or something else?

**Comment by maemigh** [comment]
Actually it's just simple text. `<pre>` monospace text with some `<br/>`s.

**Comment by torkelo** [comment]
that won't work, the new table will not support dynamic (different) table cell heights. As rows are virtualized the row height needs to be predefined and cannot differ between rows. 

But I plan to add a cell size or height option to support cells with higher content (like images)

**Comment by maemigh** [comment]
@torkelo This already works in 6.7.3 as well as in tables migrated from 6.7.3.

**Comment by torkelo** [comment]
Yes, but won’t work in 7.0 table as row height, cell height is fixed

**Comment by therealcodlord** [comment]
Personally I use the sanatize html feature of the old table panel for lots of purposes. It's one of the main reasons I started using Grafana. It makes the table panel incredibly flexible/powerful.

1) to render <img> tags so I can show images in tables. And the `Render value as link` option allows the images to be clickable to show a full-size version. 

2) to render bold/highlighting/colours/newlines etc. of certain elements within a table.

3) to show a tables with (clickable) status icons including font awesome icons etc.

**Comment by SirobbsCrafter** [comment]
We **really** used to "**Column Styles**" feature, but in **Grafana 7.0** it **faded**. We setting up column display styles **very often** and now have a problems with it. Temporary solution for that is **importing** dashboard from **old-version** of Grafana. In that case required settings tab reappearing on panel settings.

**Comment by torkelo** [comment]
You can use the Overrides tab to define something exactly similar to Column styles, where you can define cell styles, thresholds, links etc per column. 

We are missing some cell style options but that will come

**Comment by SirobbsCrafter** [comment]
@torkelo we need to format column value as date, but **Overrides** has not this feature at this time (if i not mistaking). Waiting for updates.

**Comment by wyxa** [comment]
Showing HTML in the table cells was a really powerful feature.
Combined with the variables to refer to other cells it was really flexible.
Would be nice to have it back.

**Comment by Breee** [comment]
Until this is possible: 

If you want to use the old table style,  just change the panel type to "table-old" in the dashboard's JSON.

**Comment by maxgodsk** [comment]
> Until this is possible:
> 
> You can use the old tables instead of the new ones, just change the panel type to "table-old" in the dashboard's JSON.

Nice tip! thanks

**Comment by merryllem** [comment]
Yes, Please add this feature back. 
Use cases for me:
Links, Icons etc... 
![sample use](https://user-images.githubusercontent.com/30536274/86426316-f63fe200-bcac-11ea-892f-4796af28a87e.PNG)

**Comment by patrikTSS** [comment]
I also have an issue, which also related to cell height and was possible before 7 and now it's gone.

Before 7 it was possible with "Preserve Fromatting" to display a multiline String in a cell. 
Now line breaks are igonred and there is no comparable option in the new override tab.
So basically now i have to decide whether to have horizontal scrollable headers (new table) or to have multiline Strings in the cells (old table).

If dynamic row heights are not possible, i would a appreciate it to have an option to preserve line breaks and set a fixed height for rows.

**Comment by purpleheadqueeflicker** [comment]
i have many use cases for this feature, also it will be great if we can change the style (if i want to change the colors of the links based on my own logic ). grafana 6 sanitize is removing "style" labels.

**Comment by waterdrop01** [comment]
@torkelo Is there a solution to display an **image** as the cell display mode?
![Screenshot from 2020-09-22 11-03-41](https://user-images.githubusercontent.com/47316824/93863059-500be300-fcc3-11ea-942b-1b6ccdcc6166.png)
The column would contain a public URL of an image, that would get displayed in the table.
Would be _very_ useful...

**Comment by torkelo** [comment]
image cell mode you ask for? 
well had a bit of fun this weekend 
https://github.com/grafana/grafana/pull/27669

**Comment by sulliwane** [comment]
@torkelo wow, looks insane!! Can't wait for it :)

**Comment by waterdrop01** [comment]
@torkelo wow, looks insane!! Can't wait for it :)

**Comment by Trovalo** [comment]
@torkelo I was using the "Preserve Formatting" options which is not present in v7.2

Basically, my use case was to use the table as a huge cell to show formatted text. Is this going to be supported somehow in the new table?

![image](https://user-images.githubusercontent.com/12230314/95959726-24cc6d80-0e03-11eb-9c5a-81d62669ce37.png)

**Comment by torkelo** [comment]
Don't think so, maybe in the text panel?

**Comment by Trovalo** [comment]
I can't use a query in the text panel, that's why I'm using the table
(the "Statement Text" is the result of a query, I ensure it returns only one row)

**Comment by torkelo** [comment]
but what if you could?

**Comment by Trovalo** [comment]
Oh well, that would be amazing, I'll switch to the Text Panel if that was the case.

It would be handy to have some basic transformation, in my case just the "Organize Fields" suffice.
I'm using InfluxDB, and the time column is always returned even if you don't need it, so I usually set it as hidden.

In general, I think being able to fetch text using a query would be really handy, I'm thinking about getting log details and similar use cases.

**Comment by cicciopizza** [comment]
> Showing HTML in the table cells was a really powerful feature.
> Combined with the variables to refer to other cells it was really flexible.
> Would be nice to have it back.

News about this feature ??

**Comment by wyxa** [comment]
@torkelo this was one of the most powerful features of the old table
I was really surprised that it wasn't included in the new table

**Comment by havedill** [comment]
Preserve formatting is also one that is missing from the V7 Table, even in the overrides.

**Comment by cicciopizza** [comment]
I have a lot of dashes where i use html in cells.
the HTML is generated directly from the output of the query (I created some views in mysql where some columns have the html ready to be displayed)
For example, based on the name I get the color of the text that I see in the table:
in the view I already have the html ready

```
<span><b><font color="#0082F6">ntcy_raspicy_plc</font></b></span>
<span><b><font color="#FF8A00">ntcy-raspicy-pro4</font></b></span>
.....
.....
```
and in grafana table I just need to view that column of the DB to get the text with the font / color etc. that I want.

![Selezione_037](https://user-images.githubusercontent.com/26459829/99151519-1e4c4280-269c-11eb-8541-b74e925f3418.png)

how could i get this in grafana 7.x ?

**Comment by havedill** [comment]
My example would be with the Zabbix plugin. The text metrics it pulls, (SNMP Traps for example) and multiline, and using the new table converts it all into a single line, making it unreadable. 

I also just tried out the 7.3 grafana, and that appears to break table-old entirely for me

**Comment by ralequi** [comment]
Hello,

Are there any advance on this feature?
We are stuck on Grafana 7.2 using table_old and wont upgrade (as previous comment says it is broken on 7.3)

Our use for HTML in tables is mostly to work with external links.

Thank you!

**Comment by Elufimov** [comment]
We also use table as @Trovalo in https://github.com/grafana/grafana/issues/24482#issuecomment-708230159

**Comment by maemigh** [comment]
@torkelo In additional to sanitizing, would it be possible to add an option to explicitly trust HTML like in this [blog post?](https://odetocode.com/blogs/scott/archive/2014/09/10/a-journey-with-trusted-html-in-angularjs.aspx)

This should provide even more flexibility with better style tags, custom JS, etc.

**Comment by krswin** [comment]
when html render will be supported in table ? thanks

**Comment by cicciopizza** [comment]
Hope they add this feature soon. 
I am stuck at 6.3.5 due to this lack :(

**Comment by krswin** [comment]
@cicciopizza in 6.3.5 is <span style="color: green; font-weight: bold;"> </span> works?

**Comment by cicciopizza** [comment]
> @cicciopizza in 6.3.5 is `<span style="color: green; font-weight: bold;"> </span>` works?

Yes in 6.3.5 this work for me:
<span><b><font color="#0082F6">xxxxxxx</font></b></span>

**Comment by krswin** [comment]
> > @cicciopizza in 6.3.5 is `<span style="color: green; font-weight: bold;"> Test 123 </span>` works?
> 
> Yes in 6.3.5 this work for me:
> xxxxxxx

i have tried with 6.3.5 but its hiding the span tag and there is not color visible. Can you please share which configuration do you use? thanks

**Comment by cicciopizza** [comment]
> i have tried with 6.3.5 but its hiding the span tag and there is no color or bold visible. Can you please share which configuration do you use? thanks

Did you set the 'Sanitize Html' flag in Column Style?
![Selezione_002](https://user-images.githubusercontent.com/26459829/105350670-cd9d1c80-5beb-11eb-85d3-19f52ed7d310.png)

**Comment by krswin** [comment]
Yes i have set Sanitize Html flag. `<b>test123</b>` works fine problem is with color: green with style.

**Comment by cicciopizza** [comment]
> Yes i have set Sanitize Html flag. `<b>test123</b>` works fine but problem is with color: green with style do not work. So i am not sure what is your setting so it works for you.
> 
> Do you have any other settings, other than your screenshot that below works as Html.
> `<span style="color: green; font-weight: bold;"> Test 123 </span> `

I use:
<span><b><font color="#21DF21">Test 123</font></b></span>

**Comment by krswin** [comment]
> Until this is possible:
> 
> You can use the old tables instead of the new ones, just change the panel type to "table-old" in the dashboard's JSON.
@Breee 
I have tried this but this do not work. Can you please tell what kind of settings exactly do you have? Thanks.

**Comment by cicciopizza** [comment]
> > Until this is possible:
> > You can use the old tables instead of the new ones, just change the panel type to "table-old" in the dashboard's JSON.
> > @Breee
> > I have tried this but this do not work. Can you please tell what kind of settings exactly do you have? Thanks.


I created the table in grafana 6.3.5 then I exported the dashboard and imported it in grafana 7

**Comment by Breee** [comment]
here, just set "type" to  "table-old" in the panel json.

```json
{
  "columns": [],
  "datasource": "Some Dashboard",
  "fieldConfig": {
    "defaults": {
      "custom": {}
    },
    "overrides": []
  },
  "fontSize": "90%",
  "gridPos": {
    "h": 15,
    "w": 9,
    "x": 0,
    "y": 0
  },
  "id": 2,
  "repeatDirection": "h",
  "scroll": true,
  "showHeader": true,
  "sort": {
    "col": 4,
    "desc": false
  },
  "styles": [
    {
      "$$hashKey": "object:465",
      "alias": "Build",
      "align": "auto",
      "colorMode": null,
      "colors": [
        "rgba(245, 54, 54, 0.9)",
        "rgba(237, 129, 40, 0.89)",
        "rgba(50, 172, 45, 0.97)"
      ],
      "dateFormat": "YYYY-MM-DD HH:mm:ss",
      "decimals": 2,
      "link": true,
      "linkTargetBlank": true,
      "linkTooltip": "Go to build definition",
      "linkUrl": "http://someurl/${__cell_2}/_build?${__cell}",
      "mappingType": 1,
      "pattern": "definition",
      "thresholds": [],
      "type": "string",
      "unit": "short"
    }
  ],
  "targets": [
    {
      "alias": "$tag_teamproject $tag_definition",
      "groupBy": [
        {
          "params": [
            "definition"
          ],
          "type": "tag"
        },
        {
          "params": [
            "teamproject"
          ],
          "type": "tag"
        }
      ],
      "hide": false,
      "measurement": "TfsBuilds_tfsmonitoring",
      "orderByTime": "ASC",
      "policy": "default",
      "query": "xxxxxxxxx",
      "rawQuery": false,
      "refId": "B",
      "resultFormat": "table",
      "select": [
        [
          {
            "params": [
              "seconds_since_last_build_finished"
            ],
            "type": "field"
          },
          {
            "params": [],
            "type": "last"
          },
          {
            "params": [
              "Last build"
            ],
            "type": "alias"
          }
        ],
        [
          {
            "params": [
              "seconds_since_last_failed_build_finished"
            ],
            "type": "field"
          },
          {
            "params": [
              "Failed"
            ],
            "type": "alias"
          }
        ]
      ],
      "tags": []
    }
  ],
  "title": "Build results",
  "transform": "table",
  "type": "table-old",
  "pageSize": null,
  "repeat": null,
  "timeFrom": null,
  "timeShift": null
}
```

**Comment by torkelo** [comment]
This should easy for anyone who wants to contribute, the cell display modes are easy to add more of.

**Comment by krswin** [comment]
> > > Until this is possible:
> > > You can use the old tables instead of the new ones, just change the panel type to "table-old" in the dashboard's JSON.
> > > @Breee
> > > I have tried this but this do not work. Can you please tell what kind of settings exactly do you have? Thanks.
> 
> I created the table in grafana 6.3.5 then I exported the dashboard and imported it in grafana 7

@Breee thank you. This works.

**Comment by krswin** [comment]
> Hope they add this feature soon.
> I am stuck at 6.3.5 due to this lack :(

@cicciopizza please try solution from @Breee. It worked for me so you can switch to Grafana 7.

**Comment by cicciopizza** [comment]
> > Hope they add this feature soon.
> > I am stuck at 6.3.5 due to this lack :(
> 
> @cicciopizza please try solution from @Breee. It worked for me so you can switch to Grafana 7.

@krswin yes @Breee's solution works too (I had already tried it) ...

But I often change the table, and I would be forced to change it in 6.3.5 and then do export / import again

This is the shortest way for me ... 
But it is still complicated as I would have to have a 6.3.5 graph to do this...

**Comment by lux4rd0** [comment]
I'm looking forward to being able to use sanitized HTML in a table for the following use case:

<img width="1380" alt="loki_table_override" src="https://user-images.githubusercontent.com/30187533/111037691-e3a8ab80-83ea-11eb-9e3a-ea986424ec40.png">

I'm pulling in real-time streaming and forecast weather JSON logs from an IoT device and dropping them into Loki. There are icon definitions available that I use as labels that can use an Override (along with several transformations) to then show an Icon (or even better - Base64 to locally host).

I know this is a really specific use case - but I tried using the old 6.x table visualization and it doesn't have enough overrides to get extracted JSON Loki data looking right. I'll wait and see what happens though!!

_(Would also be really nice to have an epoch time unit that doesn't require milliseconds to function as Loki data sources don't have math functions built into extracted JSON labels yet to do the * 1000...)_

**Comment by RishoT17** [comment]
I want to use this feature (HTML formatting) to show dynamic links and icons. By dynamic, I mean each cell to have a different link, or to point to a different dashboard (not just a different variable, but actually a different dashboard). 

For example:
I have a table of auto-tests performed. There can be 3-4 different test types, each having their own specialized dashboard. In this table, I want to show all test results in the same table, so that they can all be filtered and sorted together and the last cell of each row would be a "More details" link to open the specialized dashboard for that given test, based on its type.

This is currently not possible (as far as I know). 

If this was possible in the past and it got removed, I'm starting to worry that some of the existing features that we are currently using might get removed in the future too.

**Comment by Jojal** [comment]
Same problem for me. It's really important to display html in table :( . What is the status of this issue ?

**Comment by cicciopizza** [comment]
I have almost resigned myself to the fact that they will never implement this functionality ... 
I am using the old table until it works. 

The new table is really  unsuitable, it also has other problems ...

> Same problem for me. It's really important to display html in table :( . What is the status of this issue ?

**Comment by solaris-9** [comment]
has this issue been addressed?

**Comment by cicciopizza** [comment]
> has this issue been addressed?

The only hope is to keep this request active, hoping in this way that they will consider adding this fundamental feature :(

**Comment by yulipumal** [comment]
Also in my case:
Sanitized HTML in tables is a heavily used feature in my use case. Is is used especially for images in cells. Please bring it back to the new table. Running on "table-old" won't be supported for a long time and the HTML support is a strong feature for all users.

**Comment by codeincarnate** [comment]
This is something we're looking into. As per the above discussion the virtualized rendering of `react-table` makes this tricky, but is important from a performance perspective. We're working through a set of changes and tracking those here #56220

**Comment by yulipumal** [comment]
@cicciopizza @Jojal @outtersg @RishoT17 @lux4rd0 @merryllem-zz


@codeincarnate is looking for use cases of sanitized HTML in tables. Can you share your usecases here?
https://github.com/grafana/grafana/issues/56220#issuecomment-1608667843

**Comment by outtersg** [comment]
@yulipumal @codeincarnate :
I'm rendering a list of incidents or defects numbers ("ITENTERP-12345, ITENTERP-56789, INC0123456") as links to the ticketing systems:
```sql
regexp_replace(regexp_replace(anos, '(ITENTERP-[0-9]+)', '<a href="https://jira.xxx.fr/browse/\1">\1</a>', 'g'), '(INC[0-9]+)', '<a href="https://xxx.service-now.com/incident.do?sysparm_query=number=\1">\1</a>', 'g')
```

So my goal is basically **multi-link cells**, each URL being dynamically computed for a comma-separated item.

**Comment by cicciopizza** [comment]
My goal is to display a table whose data source is a MySql table which contains fields with html.
The HTML code in the table contains links, colors and fonts that are dynamically generated by a script based on pre-established logics

**Comment by oscarkilhed** [comment]
@cicciopizza 
> My goal is to display a table whose data source is a MySql table which contains fields with html. The HTML code in the table contains links, colors and fonts that are dynamically generated by a script based on pre-established logics

Just be aware, that if this is implemented, it's never going to support javascript as part of the HTML. The HTML in your scenario would likely need to be generated outside of Grafana and passed in as a value through a data source.

**Comment by Breee** [comment]
Or generated in a panel that uses the dynamic Text plugin
https://grafana.com/grafana/plugins/marcusolsson-dynamictext-panel/


In Addition, for my usecases I found that datalinks are sufficient as you
can reference variables. Probably I can Post an example in the next days

**Comment by Jeroenvandepol** [comment]
Ooh, I found an old open issue that I ran into as well. Use case: I have an RSS feed that I get with an infinity datasource. In the RSS feed there is some HTML, mostly consists off <br/> <p> and <strong>. The table pannel shows the explicit HTML, instead of the translated html.

**Comment by oiercesat** [comment]
> [@yulipumal](https://github.com/yulipumal) [@codeincarnate](https://github.com/codeincarnate) : I'm rendering a list of incidents or defects numbers ("ITENTERP-12345, ITENTERP-56789, INC0123456") as links to the ticketing systems:
> 
> regexp_replace(regexp_replace(anos, '(ITENTERP-[0-9]+)', '<a href="https://jira.xxx.fr/browse/\1">\1</a>', 'g'), '(INC[0-9]+)', '<a href="https://xxx.service-now.com/incident.do?sysparm_query=number=\1">\1</a>', 'g')
> So my goal is basically **multi-link cells**, each URL being dynamically computed for a comma-separated item.

I think this feature is needed since 4 years, and someone created a PR https://github.com/grafana/grafana/pull/68326, but no update or review since one or two years about this one. 
If this feature couldn’t be implemented, someone have an idea to do this kind of content in a cell ?
